File: src/Java/myservicebus-rabbitmq/src/test/java/com/myservicebus/rabbitmq/RabbitMqFactoryConfiguratorTests.java
package com.myservicebus.rabbitmq;

import static org.junit.jupiter.api.Assertions.*;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.atomic.AtomicReference;
import java.lang.reflect.Field;
import java.util.function.Function;

import org.junit.jupiter.api.Test;

import com.myservicebus.*;
import com.myservicebus.MessageEntityNameFormatterSpecific;
import com.myservicebus.topology.*;
import com.myservicebus.di.ServiceCollection;
import com.myservicebus.di.ServiceProvider;

public class RabbitMqFactoryConfiguratorTests {
    static class MyMessage {
    }

    static class MyConsumer implements Consumer<MyMessage> {
        @Override
        public CompletableFuture<Void> consume(ConsumeContext<MyMessage> context) {
            return CompletableFuture.completedFuture(null);
        }
    }

    interface ExternalDependency {
    }

    static class ExternalConsumer implements Consumer<MyMessage> {
        ExternalConsumer(ExternalDependency dependency) {
        }

        @Override
        public CompletableFuture<Void> consume(ConsumeContext<MyMessage> context) {
            return CompletableFuture.completedFuture(null);
        }
    }

    @MessageConsumer("attribute-orders")
    static class ExplicitEndpointConsumer implements Consumer<MyMessage> {
        @Override
        public CompletableFuture<Void> consume(ConsumeContext<MyMessage> context) {
            return CompletableFuture.completedFuture(null);
        }
    }

    @Test
    public void factoryConfiguresNonDefaultPort() {
        RabbitMqFactoryConfigurator configurator = new RabbitMqFactoryConfigurator();

        configurator.host("container-host", 32789);

        assertEquals("container-host", configurator.getClientHost());
        assertEquals(32789, configurator.getClientPort());
    }

    @Test
    public void factoryRegistersConsumerWithoutBindingItToTheInternalProvider() throws Exception {
        AtomicReference<Class<?>> requestedConsumerType = new AtomicReference<>();
        RabbitMqFactoryConfigurator configurator = new RabbitMqFactoryConfigurator();
        configurator.message(MyMessage.class, message -> message.setEntityName("external-message"));
        configurator.setConsumerFactory((ignored, consumerType) -> {
            requestedConsumerType.set(consumerType);
            return new ConsumerFactory() {
                @Override
                public <TConsumer, T> CompletableFuture<Void> send(
                        Class<TConsumer> type,
                        ConsumeContext<T> context,
                        Pipe<ConsumerConsumeContext<TConsumer, T>> next) {
                    return CompletableFuture.completedFuture(null);
                }
            };
        });
        configurator.receiveEndpoint("external-orders", endpoint -> {
            endpoint.prefetchCount(12);
            endpoint.concurrentMessageLimit(4);
            endpoint.setQueueArgument("x-queue-type", "quorum");
            endpoint.consumer(MyMessage.class, ExternalConsumer.class);
        });

        ServiceCollection services = ServiceCollection.create();
        configurator.configure(services);
        ServiceProvider provider = services.buildServiceProvider();
        MessageBusImpl bus = (MessageBusImpl) provider.getService(MessageBus.class);
        ConsumerTopology definition = provider.getService(TopologyRegistry.class).getConsumers().stream()
                .filter(candidate -> candidate.getConsumerType().equals(ExternalConsumer.class))
                .findFirst()
                .orElseThrow();

        assertEquals("external-orders", definition.getQueueName());
        assertEquals("external-message", definition.getBindings().get(0).getEntityName());
        assertEquals(12, definition.getPrefetchCount());
        assertEquals(4, definition.getConcurrentMessageLimit());
        assertEquals("quorum", definition.getQueueArguments().get("x-queue-type"));

        Field factoryField = MessageBusImpl.class.getDeclaredField("consumerFactoryFactory");
        factoryField.setAccessible(true);
        @SuppressWarnings("unchecked")
        Function<Class<?>, ConsumerFactory> factory =
                (Function<Class<?>, ConsumerFactory>) factoryField.get(bus);
        factory.apply(ExternalConsumer.class);

        assertEquals(ExternalConsumer.class, requestedConsumerType.get());
    }

    @Test
    public void consumerDefinitionReflectsCustomQueueAndExchange() {
        ServiceCollection services = ServiceCollection.create();
        BusRegistrationConfiguratorImpl cfg = new BusRegistrationConfiguratorImpl(services);
        cfg.addConsumer(MyConsumer.class);

        RabbitMqTransport.configure(cfg);
        cfg.complete();

        ServiceProvider provider = services.buildServiceProvider();
        BusRegistrationContext context = new BusRegistrationContext(provider);
        RabbitMqFactoryConfigurator factoryConfigurator = provider.getService(RabbitMqFactoryConfigurator.class);

        factoryConfigurator.message(MyMessage.class, m -> m.setEntityName("custom-exchange"));

        factoryConfigurator.receiveEndpoint("custom-queue", e -> {
            e.configureConsumer(context, MyConsumer.class);
        });

        TopologyRegistry registry = provider.getService(TopologyRegistry.class);
        ConsumerTopology def = registry.getConsumers().stream()
                .filter(d -> d.getConsumerType().equals(MyConsumer.class))
                .findFirst()
                .orElseThrow();

        assertEquals("custom-queue", def.getQueueName());
        assertEquals("custom-exchange", def.getBindings().get(0).getEntityName());
    }

    static class StaticFormatter<T> implements MessageEntityNameFormatterSpecific<T> {
        @Override
        public String formatEntityName() {
            return "formatted-" + MyMessage.class.getSimpleName().toLowerCase();
        }
    }

    @Test
    public void messageUsesEntityNameFormatter() {
        ServiceCollection services = ServiceCollection.create();
        BusRegistrationConfiguratorImpl cfg = new BusRegistrationConfiguratorImpl(services);
        cfg.addConsumer(MyConsumer.class);

        RabbitMqTransport.configure(cfg);
        cfg.complete();

        ServiceProvider provider = services.buildServiceProvider();
        BusRegistrationContext context = new BusRegistrationContext(provider);
        RabbitMqFactoryConfigurator factoryConfigurator = provider.getService(RabbitMqFactoryConfigurator.class);

        factoryConfigurator.message(MyMessage.class, m -> m.setEntityNameFormatter(new StaticFormatter<>()));
        factoryConfigurator.receiveEndpoint("custom-queue", e -> e.configureConsumer(context, MyConsumer.class));

        TopologyRegistry registry = provider.getService(TopologyRegistry.class);
        ConsumerTopology def = registry.getConsumers().stream()
                .filter(d -> d.getConsumerType().equals(MyConsumer.class))
                .findFirst()
                .orElseThrow();

        assertEquals("formatted-mymessage", def.getBindings().get(0).getEntityName());
    }

    @Test
    public void receiveEndpointAddsMessageRetry() {
        ServiceCollection services = ServiceCollection.create();
        BusRegistrationConfiguratorImpl cfg = new BusRegistrationConfiguratorImpl(services);
        cfg.addConsumer(MyConsumer.class);

        RabbitMqTransport.configure(cfg);
        cfg.complete();

        ServiceProvider provider = services.buildServiceProvider();
        BusRegistrationContext context = new BusRegistrationContext(provider);
        RabbitMqFactoryConfigurator factoryConfigurator = provider.getService(RabbitMqFactoryConfigurator.class);

        factoryConfigurator.receiveEndpoint("custom-queue", e -> {
            e.useMessageRetry(r -> r.immediate(2));
            e.configureConsumer(context, MyConsumer.class);
        });

        TopologyRegistry registry = provider.getService(TopologyRegistry.class);
        ConsumerTopology def = registry.getConsumers().stream()
                .filter(d -> d.getConsumerType().equals(MyConsumer.class))
                .findFirst()
                .orElseThrow();

        assertNotNull(def.getConfigure());
    }

    @Test
    public void receiveEndpointSetsQueueArguments() {
        ServiceCollection services = ServiceCollection.create();
        BusRegistrationConfiguratorImpl cfg = new BusRegistrationConfiguratorImpl(services);
        cfg.addConsumer(MyConsumer.class);

        RabbitMqTransport.configure(cfg);
        cfg.complete();

        ServiceProvider provider = services.buildServiceProvider();
        BusRegistrationContext context = new BusRegistrationContext(provider);
        RabbitMqFactoryConfigurator factoryConfigurator = provider.getService(RabbitMqFactoryConfigurator.class);

        factoryConfigurator.receiveEndpoint("custom-queue", e -> {
            e.setQueueArgument("x-queue-type", "quorum");
            e.configureConsumer(context, MyConsumer.class);
        });

        TopologyRegistry registry = provider.getService(TopologyRegistry.class);
        ConsumerTopology def = registry.getConsumers().stream()
                .filter(d -> d.getConsumerType().equals(MyConsumer.class))
                .findFirst()
                .orElseThrow();

        assertEquals("quorum", def.getQueueArguments().get("x-queue-type"));
    }

    @Test
    public void configureEndpointsUsesFormatter() {
        ServiceCollection services = ServiceCollection.create();
        BusRegistrationConfiguratorImpl cfg = new BusRegistrationConfiguratorImpl(services);
        cfg.addConsumer(MyConsumer.class);

        RabbitMqTransport.configure(cfg);
        cfg.complete();

        ServiceProvider provider = services.buildServiceProvider();
        BusRegistrationContext context = new BusRegistrationContext(provider);
        RabbitMqFactoryConfigurator factoryConfigurator = provider.getService(RabbitMqFactoryConfigurator.class);

        factoryConfigurator.setEndpointNameFormatter(mt -> "formatted-" + mt.getSimpleName().toLowerCase());
        factoryConfigurator.configureEndpoints(context);

        TopologyRegistry registry = provider.getService(TopologyRegistry.class);
        ConsumerTopology def = registry.getConsumers().stream()
                .filter(d -> d.getConsumerType().equals(MyConsumer.class))
                .findFirst()
                .orElseThrow();

        assertEquals("formatted-myconsumer", def.getQueueName());
    }

    @Test
    public void configureEndpointsDoesNotReplaceExplicitEndpoint() {
        ServiceCollection services = ServiceCollection.create();
        BusRegistrationConfiguratorImpl cfg = new BusRegistrationConfiguratorImpl(services);
        cfg.addConsumer(ExplicitEndpointConsumer.class);

        RabbitMqTransport.configure(cfg);
        cfg.complete();

        ServiceProvider provider = services.buildServiceProvider();
        BusRegistrationContext context = new BusRegistrationContext(provider);
        RabbitMqFactoryConfigurator factoryConfigurator = provider.getService(RabbitMqFactoryConfigurator.class);

        factoryConfigurator.setEndpointNameFormatter(mt -> "formatted-" + mt.getSimpleName().toLowerCase());
        factoryConfigurator.configureEndpoints(context);

        TopologyRegistry registry = provider.getService(TopologyRegistry.class);
        ConsumerTopology definition = registry.getConsumers().stream()
                .filter(consumer -> consumer.getConsumerType().equals(ExplicitEndpointConsumer.class))
                .findFirst()
                .orElseThrow();

        assertEquals("attribute-orders", definition.getQueueName());
    }
}

File: test/MyServiceBus.RabbitMq.Tests/RabbitMqFactoryConfiguratorTests.cs
using Microsoft.Extensions.DependencyInjection;
using MyServiceBus;
using MyServiceBus.Topology;
using MyServiceBus.Serialization;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using Xunit;
using Xunit.Sdk;

namespace MyServiceBus.RabbitMq.Tests;

public class RabbitMqFactoryConfiguratorTests
{
    class MyMessage { }

    class MyConsumer : IConsumer<MyMessage>
    {
        public Task Consume(ConsumeContext<MyMessage> context) => Task.CompletedTask;
    }

    class CustomConsumerFactory<TConsumer> : IConsumerFactory<TConsumer>
        where TConsumer : class
    {
        public Task Send<TMessage>(
            ConsumeContext<TMessage> context,
            IPipe<ConsumerConsumeContext<TConsumer, TMessage>> next)
            where TMessage : class => Task.CompletedTask;
    }

    class TestMessageBus : IMessageBus
    {
        public ConsumerTopology? AddedConsumer { get; private set; }
        public Uri Address => new("loopback://localhost/");
        public IBusTopology Topology => new TopologyRegistry();
        public Task StartAsync(CancellationToken cancellationToken) => Task.CompletedTask;
        public Task StopAsync(CancellationToken cancellationToken) => Task.CompletedTask;
        public Task Publish<T>(object message, Action<IPublishContext>? contextCallback = null, CancellationToken cancellationToken = default) where T : class => Task.CompletedTask;
        public Task Publish<T>(T message, Action<IPublishContext>? contextCallback = null, CancellationToken cancellationToken = default) where T : class => Task.CompletedTask;
        public IPublishEndpoint GetPublishEndpoint() => this;
        public Task<ISendEndpoint> GetSendEndpoint(Uri uri) => Task.FromResult<ISendEndpoint>(new StubSendEndpoint());
        public Task AddConsumer<TMessage, TConsumer>(ConsumerTopology consumer, Delegate? configure = null, CancellationToken cancellationToken = default)
            where TConsumer : class, IConsumer<TMessage>
            where TMessage : class
        {
            AddedConsumer = consumer;
            return Task.CompletedTask;
        }

        public Task AddHandler<TMessage>(string queueName, string exchangeName, Func<ConsumeContext<TMessage>, Task> handler, int? retryCount = null, TimeSpan? retryDelay = null, ushort? prefetchCount = null, IDictionary<string, object?>? queueArguments = null, IMessageSerializer? serializer = null, CancellationToken cancellationToken = default) where TMessage : class => Task.CompletedTask;

        class StubSendEndpoint : ISendEndpoint
        {
            public Task Send<T>(T message, Action<ISendContext>? contextCallback = null, CancellationToken cancellationToken = default) where T : class => Task.CompletedTask;
            public Task Send<T>(object message, Action<ISendContext>? contextCallback = null, CancellationToken cancellationToken = default) where T : class => Task.CompletedTask;
        }
    }

    [Fact]
    public void Factory_endpoint_registers_typed_consumer_without_service_binding()
    {
        var registry = new TopologyRegistry();
        var bus = new TestMessageBus();
        var services = new ServiceCollection();
        services.AddSingleton(registry);
        services.AddSingleton<IMessageBus>(bus);
        var provider = services.BuildServiceProvider();
        var endpointActions = new List<Action<IMessageBus, IServiceProvider>>();
        var endpoint = new ReceiveEndpointConfigurator("external-orders", new Dictionary<Type, string>(), endpointActions);

        endpoint.Consumer<MyConsumer, MyMessage>();
        endpointActions.Single()(bus, provider);

        var consumer = Assert.Single(registry.Consumers);
        Assert.Equal(typeof(MyConsumer), consumer.ConsumerType);
        Assert.Equal(typeof(MyMessage), consumer.Bindings[0].MessageType);
        Assert.Equal("external-orders", consumer.QueueName);
        Assert.Same(consumer, bus.AddedConsumer);
        Assert.Null(provider.GetService<MyConsumer>());
    }

    [Fact]
    public void Factory_uses_configured_consumer_factory_type()
    {
        var configurator = new RabbitMqFactoryConfigurator();
        configurator.SetConsumerFactory(typeof(CustomConsumerFactory<>));
        var services = new ServiceCollection();

        configurator.Configure(services);
        var provider = services.BuildServiceProvider();

        Assert.IsType<CustomConsumerFactory<MyConsumer>>(
            provider.GetRequiredService<IConsumerFactory<MyConsumer>>());
    }

    class TestBusRegistrationContext : IBusRegistrationContext
    {
        public TestBusRegistrationContext(IServiceProvider provider) => ServiceProvider = provider;
        public IServiceProvider ServiceProvider { get; }
    }

    class TestRabbitMqFactoryConfigurator : IRabbitMqFactoryConfigurator
    {
        private readonly Dictionary<Type, string> _exchangeNames = new();
        public IEndpointNameFormatter? EndpointNameFormatter { get; private set; }
        public IMessageEntityNameFormatter? EntityNameFormatter { get; private set; }
        public string ClientHost => "localhost";
        public int ClientPort => 5672;
        public ushort PrefetchCount { get; private set; }

        public void Message<T>(Action<MessageConfigurator> configure)
        {
            configure(new MessageConfigurator(typeof(T), _exchangeNames));
        }

        public void ReceiveEndpoint(string queueName, Action<ReceiveEndpointConfigurator> configure)
        {
            configure(new ReceiveEndpointConfigurator(queueName, _exchangeNames, new List<Action<IMessageBus, IServiceProvider>>()));
        }

        public void Host(string host, Action<IRabbitMqHostConfigurator>? configure = null)
        {
        }

        public void Host(string host, int port, Action<IRabbitMqHostConfigurator>? configure = null)
        {
        }

        public void SetEndpointNameFormatter(IEndpointNameFormatter formatter)
        {
            EndpointNameFormatter = formatter;
        }

        public void SetEntityNameFormatter(IMessageEntityNameFormatter formatter)
        {
            EntityNameFormatter = formatter;
        }

        public void SetPrefetchCount(ushort prefetchCount)
        {
            PrefetchCount = prefetchCount;
        }
        public void SetConsumerFactory(Type consumerFactoryType)
        {
        }
    }

    [Fact]
    public void ReceiveEndpoint_overrides_queue_and_exchange()
    {
        var registry = new TopologyRegistry();
        registry.RegisterConsumer<MyConsumer>("original-queue", null, typeof(MyMessage));

        var services = new ServiceCollection();
        services.AddSingleton(registry);
        services.AddSingleton<IMessageBus, TestMessageBus>();
        var provider = services.BuildServiceProvider();
        var context = new TestBusRegistrationContext(provider);

        var configurator = new TestRabbitMqFactoryConfigurator();
        configurator.Message<MyMessage>((m) => m.SetEntityName("custom-exchange"));
        configurator.ReceiveEndpoint("custom-queue", (e) => e.ConfigureConsumer<MyConsumer>(context));

        var def = registry.Consumers.First(c => c.ConsumerType == typeof(MyConsumer));
        Assert.Equal("custom-queue", def.QueueName);
        Assert.Equal("custom-exchange", def.Bindings[0].EntityName);
    }

    class StaticEntityFormatter<T> : IMessageEntityNameFormatter<T>
    {
        public string FormatEntityName() => $"formatted-{typeof(T).Name.ToLowerInvariant()}";
    }

    [Fact]
    public void Message_uses_entity_name_formatter()
    {
        var registry = new TopologyRegistry();
        registry.RegisterConsumer<MyConsumer>("original-queue", null, typeof(MyMessage));

        var services = new ServiceCollection();
        services.AddSingleton(registry);
        services.AddSingleton<IMessageBus, TestMessageBus>();
        var provider = services.BuildServiceProvider();
        var context = new TestBusRegistrationContext(provider);

        var configurator = new TestRabbitMqFactoryConfigurator();
        configurator.Message<MyMessage>(m => m.SetEntityNameFormatter(new StaticEntityFormatter<MyMessage>()));
        configurator.ReceiveEndpoint("custom-queue", e => e.ConfigureConsumer<MyConsumer>(context));

        var def = registry.Consumers.First(c => c.ConsumerType == typeof(MyConsumer));
        Assert.Equal("formatted-mymessage", def.Bindings[0].EntityName);
    }

    [Fact]
    public void ReceiveEndpoint_adds_message_retry()
    {
        var registry = new TopologyRegistry();
        registry.RegisterConsumer<MyConsumer>("original-queue", null, typeof(MyMessage));

        var services = new ServiceCollection();
        services.AddSingleton(registry);
        services.AddSingleton<IMessageBus, TestMessageBus>();
        var provider = services.BuildServiceProvider();
        var context = new TestBusRegistrationContext(provider);

        var configurator = new TestRabbitMqFactoryConfigurator();
        configurator.ReceiveEndpoint("custom-queue", (e) =>
        {
            e.UseMessageRetry(r => r.Immediate(2));
            e.ConfigureConsumer<MyConsumer>(context);
        });

        var def = registry.Consumers.First(c => c.ConsumerType == typeof(MyConsumer));
        Assert.NotNull(def.ConfigurePipe);
    }

    [Fact]
    public void ReceiveEndpoint_sets_queue_arguments()
    {
        var registry = new TopologyRegistry();
        registry.RegisterConsumer<MyConsumer>("original-queue", null, typeof(MyMessage));

        var services = new ServiceCollection();
        services.AddSingleton(registry);
        services.AddSingleton<IMessageBus, TestMessageBus>();
        var provider = services.BuildServiceProvider();
        var context = new TestBusRegistrationContext(provider);

        var configurator = new TestRabbitMqFactoryConfigurator();
        configurator.ReceiveEndpoint("custom-queue", e =>
        {
            e.SetQueueArgument("x-queue-type", "quorum");
            e.ConfigureConsumer<MyConsumer>(context);
        });

        var def = registry.Consumers.First(c => c.ConsumerType == typeof(MyConsumer));
        Assert.Equal("quorum", def.QueueArguments!["x-queue-type"]);
    }

    class StaticFormatter : IEndpointNameFormatter
    {
        public string Format(Type messageType) => "formatted-" + messageType.Name.ToLowerInvariant();
    }

    [Fact]
    public void ConfigureEndpoints_uses_formatter()
    {
        var registry = new TopologyRegistry();
        registry.RegisterConsumer<MyConsumer>("original-queue", null, typeof(MyMessage));

        var services = new ServiceCollection();
        services.AddSingleton(registry);
        services.AddSingleton<IMessageBus, TestMessageBus>();
        var provider = services.BuildServiceProvider();
        var context = new TestBusRegistrationContext(provider);

        var configurator = new TestRabbitMqFactoryConfigurator();
        configurator.SetEndpointNameFormatter(new StaticFormatter());
        configurator.ConfigureEndpoints(context);

        var def = registry.Consumers.First(c => c.ConsumerType == typeof(MyConsumer));
        Assert.Equal("formatted-myconsumer", def.QueueName);
    }

    [Fact]
    public void ConfigureEndpoints_does_not_replace_an_explicit_endpoint()
    {
        var registry = new TopologyRegistry();
        registry.RegisterConsumer<MyConsumer, MyMessage>(
            "attribute-orders",
            endpointNameIsExplicit: true);

        var services = new ServiceCollection();
        services.AddSingleton(registry);
        services.AddSingleton<IMessageBus, TestMessageBus>();
        var provider = services.BuildServiceProvider();
        var context = new TestBusRegistrationContext(provider);

        var configurator = new TestRabbitMqFactoryConfigurator();
        configurator.SetEndpointNameFormatter(new StaticFormatter());
        configurator.ConfigureEndpoints(context);

        var definition = Assert.Single(registry.Consumers);
        Assert.Equal("attribute-orders", definition.QueueName);
    }
}
